### ISSUE TITLE
Don't require that all unit symbols shall be usable with unit prefixes

### ISSUE THREAD

**Issue by henrikt-ma** [modelica/ModelicaSpecification#3053]
Currently, if a tool allows the user to introduce the unit symbol `"in"` for the inch, then it currently follows that one may use this unit symbol with prefixes, as in `"min"` (as far as I can tell, `"min"` isn't one of the unit symbols defined by the standard, even though it's used by the MSL).  Similarly, the presence of `"degF"` in the MSL probably means that tools are also expected to recognize `"MdegF"`.  Neither of these combinations of unit prefix and unit symbol seems desirable, suggesting that we make two changes:
- Say that _some_ unit symbols can be combined with a unit prefix. (It follows that a user-defined unit must opt-in if it wants to be combinable with prefixes.)
- Clearly define which of the standard unit symbols that can be combined with a unit prefix.

See also ongoing work on #3020.

**Comment by HansOlsson** [comment]
The specification says that unit is preferred over prefix+unit; and `"min"` is defined in the SI-standard(*) as an abbreviation of minute (meaning 60 seconds - the 60th of a degree for an angle is sort of abbreviated `′` - but usually written together with degree in a way that isn't supported in Modelica).

*: Technically minute, hour, day, astronomical unit, tonne, litre, hectare etc are acceptable non-SI units according to the SI-standard.

I agree that we could change it to SI-units and optionally for other units - but based on the precedence given above it doesn't seem that urgent.

**Comment by henrikt-ma** [comment]
The urgency of it comes from us doing work in the area of unit string processing and user customization of units, and not being able to give user control over whether a unit can be used with prefixes is an annoying limitation.  Also, as the need for this is rather clear already today, it would be much better for us if we could start integrating the option into our framework now, instead of having to introduce this later when it will cause more pain.

**Comment by henrikt-ma** [comment]
> The specification says that unit is preferred over prefix+unit; and `"min"` is defined in the SI-standard(*) as an abbreviation of minute (meaning 60 seconds - the 60th of a degree for an angle is sort of abbreviated `′` - but usually written together with degree in a way that isn't supported in Modelica).
> 
> *: Technically minute, hour, day, astronomical unit, tonne, litre, hectare etc are acceptable non-SI units according to the SI-standard.

Right, `"min"` even seems to be _the_ symbol for minute (as opposed to just one of many acceptable forms):
https://en.wikipedia.org/wiki/Non-SI_units_mentioned_in_the_SI

This means that the Modelica specification needs to be clear on the status of these acceptable non-SI units, and precisely by which symbols they should be recognized.  For example, what would be the status of `"L"` for liter i Modelica?  What are the Modelica symbols for plane unit minute and second?

I also have the feeling that there are some of these non-SI units that aren't supposed to be used with prefixes.  Or how about expressing one's length as 12.2 pau?  On the other hand, there are also non-SI units that should be usable with prefixes, such as `"mL"` (or `"ml"`) for milliliter, again showing that precise control over which units that are usable with prefixes is needed.

**Comment by HansOlsson** [comment]
> This means that the Modelica specification needs to be clear on the status of these acceptable non-SI units, and precisely by which symbols they should be recognized. For example, what would be the status of `"L"` for liter i Modelica? What are the Modelica symbols for plane unit minute and second?

The problem with plane minute and second is that they are used combined with degrees, and that capability is currently lacking and I don't see a large need for it. (Writing 2 degrees 37 minutes 12 seconds.)

> I also have the feeling that there are some of these non-SI units that aren't supposed to be used with prefixes. Or how about expressing one's length as 12.2 pau? On the other hand, there are also non-SI units that should be usable with prefixes, such as `"mL"` (or `"ml"`) for milliliter, again showing that precise control over which units that are usable with prefixes is needed.

I'm not sure if there's a rigid rule. To me it's more that those units are multiples of proper SI-units because the SI-unit was not appropriate in size in that field and you then want to continue going away from the proper SI-unit. That's why we have "dl", "ml", "nl" but not "kl"; and on the other hand "tonne", "kilotonne", "gigatonne" (like for CO2 emissions), but not "millitonne".

Astronomical unit is only used inside solar systems so there's seems to be no major need for multiples.
Elsewhere in astronomy parsec is used (and there are larger multiples as well); and technically defined in terms of au.

**Comment by HansOlsson** [comment]
If we want to add something for this we could after
"A unit-operand should first be interpreted as a unit-symbol and only if not successful the second alternative assuming a prefixed operand should be exploited."
add something like:
In practice the prefixed operand are not used for all units.

**Comment by henrikt-ma** [comment]
That wouldn't solve the problem that one cannot know what the available unit symbols are on another user's setup.  By allowing a unit symbol to opt out from use with prefixes, one can also prevent bad uses of prefixes that one simply doesn't want to see in one's libraries.

**Comment by HansOlsson** [comment]
> That wouldn't solve the problem that one cannot know what the available unit symbols are on another user's setup. By allowing a unit symbol to opt out from use with prefixes, one can also prevent bad uses of prefixes that one simply doesn't want to see in one's libraries.

What is your proposed solution? We cannot give a long required list of allowed unit symbols combined unit prefixes.

**Comment by henrikt-ma** [comment]
> What is your proposed solution? We cannot give a long required list of allowed unit symbols combined unit prefixes.

Oh, sorry if I didn't make that clear.  The proposed solution is:
- Specify that a unit symbol may optionally be usable with prefixes (and then all of the prefixes).
- State which of the standard Modelica unit symbols that are usable with prefixes.
- Mention that it is up to the tools that allow for custom units to give control over which of these that may be used with prefix.  (In the extreme cases, a tool would either allow prefixes for all custom units, or for none of them.)

**Comment by HansOlsson** [comment]
> > What is your proposed solution? We cannot give a long required list of allowed unit symbols combined unit prefixes.
> 
> Oh, sorry if I didn't make that clear. The proposed solution is:
> 
> * Specify that a unit symbol may optionally be usable with prefixes (and then all of the prefixes).
> * State which of the standard Modelica unit symbols that are usable with prefixes.
> * Mention that it is up to the tools that allow for custom units to give control over which of these that may be used with prefix.  (In the extreme cases, a tool would either allow prefixes for all custom units, or for none of them.)

Ok, but I think we should start by stating something that I could find that is important for this - and then update the 2nd item based on this:
- The unit should (so there can be exceptions) use base and derived SI-units without prefixes (formulated in some good way since the base unit for mass is "kg").
- The displayUnit can use any unit-string conforming to the syntax; including any prefix for any of the SI-units. Tools may also recognize non-SI units (in particular the ones recognized by the SI-standard) and optionally allow prefixes for the non-SI units.

(Yes, I'm fully aware that the annotations do not agree with the first item.)

This means that odd prefix-combinations will only be for displayUnit and:
- Tools can assume that they are allowed if they occur; so even if someone write "apc" for "atto parsec" we could figure out the meaning.
- Tools can verify that they are consistent, so if the tool sees: `Real x(unit="s", displayUnit="min");` it is clear that it is minute. Even if the tool recognizes milli-inch, and ignores the rule that non-prefixed variant is preferred, it would see that it is inconsistent with the unit.

**Comment by henrikt-ma** [comment]
> Ok, but I think we should start by stating something that I could find that is important for this - and then update the 2nd item based on this:
> 
> * The unit should (so there can be exceptions) use base and derived SI-units without prefixes (formulated in some good way since the base unit for mass is "kg").

I think that this can be discussed separately.

> * The displayUnit can use any unit-string conforming to the syntax; including any prefix for any of the SI-units. Tools may also recognize non-SI units (in particular the ones recognized by the SI-standard) and optionally allow prefixes for the non-SI units.

It sounds a bit contradictory to both say _can use any unit-string_ and then specify restrictions, but more importantly I'm afraid I am missing the connection to the point above.

> (Yes, I'm fully aware that the annotations do not agree with the first item.)

What annotations?

> This means that odd prefix-combinations will only be for displayUnit and:

Not as long as `unit` doesn't strictly forbid the use of prefixes.

> * Tools can assume that they are allowed if they occur; so even if someone write "apc" for "atto parsec" we could figure out the meaning.

This is exactly the kind of ambiguity that I would like to avoid.  If a tool doesn't know about either the unit symbol `"apc"` or the unit symbol `"pc"` with prefixing enabled, I'd like the tool to treat this like some kind of error.  Not necessarily preventing me from translating the model, but at least treated with a warning about an unknown `displayUnit` being ignored.

> * Tools can verify that they are consistent, so if the tool sees: `Real x(unit="s", displayUnit="min");` it is clear that it is minute. Even if the tool recognizes milli-inch, and ignores the rule that non-prefixed variant is preferred, it would see that it is inconsistent with the unit.

That's too much "clever" reasoning for my taste.  To me, just using the option of having a non-prefixable `"in"` appears as a more robust approach.
